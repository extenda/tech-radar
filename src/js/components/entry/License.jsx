import React from 'react';
import PropTypes from 'prop-types';
import Markdown from 'react-markdown';
import Icon from '../Icon';

const OS_LICENSES = {
  'Apache-1.1': 'Apache Software License Version 1.1',
  'Apache-2.0': 'Apache Software License Version 2.0',
  'BSD-2': 'Simplified BSD License (BSD 2-Clause License)',
  'BSD-3': 'New BSD License (BSD 3-Clause License)',
  'BSD-4': 'Original BSD License (BSD 4-Clause License)',
  'BSD-style': 'BSD-style license, see project license for details',
  CC0: 'Creative Commons Public Domain',
  'CDDL-1.0': 'Common Development and Distribution License (CDDL) Version 1.0',
  'CCDL-1.1': 'Common Development and Distribution License (CDDL) Version 1.1',
  'EPL-1.0': 'Eclipse Public License (EPL) Version 1.0',
  'EDL-1.0': 'Eclipse Distribution License (EDL) Version 1.0',
  'LGPL-2.1': 'GNU Lesser General Public License (LGPL) Version 2.1',
  'LGPL-3.0': 'GNU Lesser General Public License (LGPL) Version 3.0',
  'GPL-CE': 'GNU General Public License (GPL) with the Classpath Exception',
  MIT: 'The MIT License',
  'MPL-1.0': 'Mozilla Public License (MPL) Version 1.0',
  'MPL-1.1': 'Mozilla Public License (MPL) Version 1.1',
  'MPL-2.0': 'Mozilla Public License (MPL) Version 2.0',
  Public: 'Public Domain',
};

const COMMERCIAL_DEFAULT_TEXT = {
  Google: 'Use of this software requires a license for Google Cloud Platform.',
};

const commercialDescription = commercial => commercial.description
  || COMMERCIAL_DEFAULT_TEXT[commercial.company];

const License = (props) => {
  const { license: { openSource, commercial } } = props;

  if (openSource == null && commercial == null) {
    return null;
  }

  return (
    <React.Fragment>
      <h2>
        License
      </h2>
      {openSource && (
        <div>
          <strong>
            <Icon name="check" />
            Available for use under an open-source license.
          </strong>
          <p>
            {openSource.link
              ? <a href={openSource.link}>{OS_LICENSES[openSource.name]}</a>
              : OS_LICENSES[openSource.name]}
          </p>
          {openSource.description && (
            <Markdown source={openSource.description} />
          )}
        </div>
      )}
      {commercial && (
        <div>
          <strong>
            <Icon name="exclamation-triangle" />
            Available for use with a commercial agreement from
            {' '}
            {commercial.company}
            .
          </strong>
          {commercialDescription(commercial) && (
            <Markdown source={commercialDescription(commercial)} />
          )}
        </div>
      )}
    </React.Fragment>
  );
};

License.propTypes = {
  license: PropTypes.shape({
    openSource: PropTypes.shape({
      name: PropTypes.string.isRequired,
      link: PropTypes.string,
      description: PropTypes.string,
    }),
    commercial: PropTypes.shape({
      company: PropTypes.string.isRequired,
      description: PropTypes.string,
    }),
  }),
};

License.defaultProps = {
  license: {
    openSource: null,
    commercial: null,
  },
};

export default License;
