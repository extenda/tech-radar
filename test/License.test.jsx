import React from 'react';
import { shallow } from 'enzyme';
import License from '../src/js/components/entry/License';

describe('<License />', () => {
  test('It renders nothing for missing license', () => {
    const component = shallow(<License />);
    expect(component).toMatchSnapshot();
  });

  test('It renders open-source license', () => {
    const component = shallow(<License license={{ openSource: { name: 'MIT' } }} />);
    expect(component).toMatchSnapshot();
    expect(component.html()).toMatch('The MIT License');
  });

  test('It renders open-source license with description', () => {
    const component = shallow(<License license={{ openSource: { name: 'MIT', description: 'OSS Desc' } }} />);
    expect(component).toMatchSnapshot();
    expect(component.html()).toMatch('The MIT License');
    expect(component.html()).toMatch('OSS Desc');
  });

  test('It renders open-source license with link', () => {
    const component = shallow(<License license={{ openSource: { name: 'MIT', link: 'https://license.com' } }} />);
    expect(component).toMatchSnapshot();
    expect(component.find('a')).toHaveLength(1);
  });

  test('It renders commercial license', () => {
    const component = shallow(<License license={{ commercial: { company: 'Oracle' } }} />);
    expect(component).toMatchSnapshot();
    expect(component.html()).toMatch('Oracle');
  });

  test('It renders commercial license with description', () => {
    const component = shallow(<License license={{ commercial: { company: 'Oracle', description: 'License text.' } }} />);
    expect(component).toMatchSnapshot();
    expect(component.html()).toMatch('License text');
  });

  test('It renders dual license', () => {
    const component = shallow(
      <License
        license={{
          commercial: { company: 'Oracle' },
          openSource: { name: 'MIT' },
        }}
      />,
    );
    expect(component).toMatchSnapshot();
    expect(component.html()).toMatch('The MIT License');
    expect(component.html()).toMatch('Oracle');
  });
});
