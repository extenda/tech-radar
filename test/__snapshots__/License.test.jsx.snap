// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<License /> It renders commercial license 1`] = `
<Fragment>
  <h2>
    License
  </h2>
  <div>
    <strong>
      <Icon
        name="exclamation-triangle"
        spaceBefore={true}
      />
      Available for use with a commercial agreement from
       
      Oracle
      .
    </strong>
  </div>
</Fragment>
`;

exports[`<License /> It renders commercial license with description 1`] = `
<Fragment>
  <h2>
    License
  </h2>
  <div>
    <strong>
      <Icon
        name="exclamation-triangle"
        spaceBefore={true}
      />
      Available for use with a commercial agreement from
       
      Oracle
      .
    </strong>
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="License text."
      sourcePos={false}
      transformLinkUri={[Function]}
    />
  </div>
</Fragment>
`;

exports[`<License /> It renders dual license 1`] = `
<Fragment>
  <h2>
    License
  </h2>
  <div>
    <strong>
      <Icon
        name="check"
        spaceBefore={true}
      />
      Available for use under an open-source license.
    </strong>
    <p>
      The MIT License
    </p>
  </div>
  <div>
    <strong>
      <Icon
        name="exclamation-triangle"
        spaceBefore={true}
      />
      Available for use with a commercial agreement from
       
      Oracle
      .
    </strong>
  </div>
</Fragment>
`;

exports[`<License /> It renders nothing for missing license 1`] = `""`;

exports[`<License /> It renders open-source license 1`] = `
<Fragment>
  <h2>
    License
  </h2>
  <div>
    <strong>
      <Icon
        name="check"
        spaceBefore={true}
      />
      Available for use under an open-source license.
    </strong>
    <p>
      The MIT License
    </p>
  </div>
</Fragment>
`;

exports[`<License /> It renders open-source license with description 1`] = `
<Fragment>
  <h2>
    License
  </h2>
  <div>
    <strong>
      <Icon
        name="check"
        spaceBefore={true}
      />
      Available for use under an open-source license.
    </strong>
    <p>
      The MIT License
    </p>
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="OSS Desc"
      sourcePos={false}
      transformLinkUri={[Function]}
    />
  </div>
</Fragment>
`;

exports[`<License /> It renders open-source license with link 1`] = `
<Fragment>
  <h2>
    License
  </h2>
  <div>
    <strong>
      <Icon
        name="check"
        spaceBefore={true}
      />
      Available for use under an open-source license.
    </strong>
    <p>
      <a
        href="https://license.com"
      >
        The MIT License
      </a>
    </p>
  </div>
</Fragment>
`;
